modyn/model_storage/internal/model_storage_manager.py

```diff
@@ -0,0 +1,190 @@
+        if not model.parent_model:
+            # base case: we can load a fully stored model
+            model_storage_strategy.full_model_strategy.load_model(model_state, self._storage_dir / model.model_path)
+            return
+
+        # recursive step: we recurse to load the model state of the parent model
+        self._reconstruct_model(model.parent_model, model_state, model_storage_strategy)
+
+        # we apply the incremental strategy to load our model state
+        model_storage_strategy.incremental_model_strategy.load_model(model_state, self._storage_dir / model.model_path)
+
+    def _get_previous_model(self, pipeline_id: int, trigger_id: int) -> Optional[TrainedModel]:
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            return (
+                database.session.query(TrainedModel)
+                .filter(TrainedModel.pipeline_id == pipeline_id, TrainedModel.trigger_id == trigger_id - 1)
+                .first()
+            )
+
+    def load_model(self, model_id: int, metadata: bool) -> Optional[dict]:
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model: Optional[TrainedModel] = database.session.get(TrainedModel, model_id)
+            if model is None:
+                logger.error(f"Model {model_id} does not exist.")
+                return None
+        model_storage_strategy = self.get_model_storage_strategy(model.pipeline_id)
+
+        model_state = self._get_base_model_state(model.pipeline_id)
+        self._reconstruct_model(model_id, model_state, model_storage_strategy)
+        model_dict = {"model": model_state}
+
+        if metadata:
+            if not model.metadata_path:
+                logger.error(f"Metadata not available for model {model_id}")
+                return None
+            with tempfile.NamedTemporaryFile() as temp_file:
+                temp_file_path = pathlib.Path(temp_file.name)
+                unzip_file(self._storage_dir / model.metadata_path, temp_file_path)
+                metadata_dict = torch.load(temp_file_path)
+            model_dict.update(metadata_dict)
+
+        return model_dict
+
+    def delete_model(self, model_id: int) -> bool:
+        with MetadataDatabaseConnection(self._modyn_config) as database:
+            model: Optional[TrainedModel] = database.session.get(TrainedModel, model_id)
+
+            if model is None:
+                logger.error(f"Trained model {model_id} was not found.")
+                return False
+            model_storage_strategy = self.get_model_storage_strategy(model.pipeline_id)
+            child_state = self._get_base_model_state(model.pipeline_id)
+
+            child: TrainedModel
+            for child in model.children:
+                assert child.pipeline_id == model.pipeline_id, "Pipeline does not match for parent and child model"
+
+                self._reconstruct_model(child.model_id, child_state, model_storage_strategy)
```

Comment:
Reworked the deletion according to our discussion. We will check, if a model has some children. If yes, deletion is not possible. Otherwise the model can be discarded safely.